### PR TITLE
 Supporting interactive cropping of meshes

### PR DESCRIPTION
**Pull request by syncle**
VisualizerWithEditing class supports interactive cropping of meshes. 
- Adding SelectDownSample function for Triangular mesh
Keep a triangle only if all the three corresponding vertices are selected by user.
Preserves vertex ordering
- Generalize Tools/ManuallyCropPointCloud to Tools/ManuallyCropGeometry
Either point cloud and mesh is supported.

**Example of mesh cropping**:
```
./ManuallyCropGeometry --mesh [path_to_mesh_file] 
```
mesh
mesh_crop

**Example of point cloud cropping**:
```
./ManuallyCropGeometry --pointcloud [path_to_mesh_or_pointcloud_file] 
```
pointcloud
pointcloud_crop

### CODE REVIEW

File: src/Core/Geometry/DownSample.cpp [hunk 1]
```diff
@@ -103,6 +104,74 @@ std::shared_ptr<PointCloud> SelectDownSample(const PointCloud &input,
...
+                    input.vertex_normals_[i]);
+            if (has_colors) output->vertex_colors_.push_back(
+                    input.vertex_colors_[i]);
+        }
+    }
+    PrintDebug("Triangle mesh sampled from %d vertices and %d triangles to %d vertices and %d triangles.\n",
+            (int)input.vertices_.size(), (int)input.triangles_.size(),
+            (int)output->vertices_.size(), (int)output->triangles_.size());
+    return output;
+}
```
Comment (syncle) [inline src/Core/Geometry/DownSample.cpp]: This function is bit long. I have tried to figure out the most efficient way to implement an algorithm that picks faces having three selected points. Maybe the names of variables are not very concise.

Comment (syncle) [reply src/Core/Geometry/DownSample.cpp]: I am not very sure this function can be placed here. In DownSample.cpp, most of the functions are point cloud operations.

File: src/Tools/CMakeLists.txt [hunk 2]
```diff
@@ -13,6 +13,6 @@ endmacro(TOOL)
 
 TOOL(ConvertPointCloud      ${CMAKE_PROJECT_NAME})
 TOOL(EncodeShader)
-TOOL(ManuallyCropPointCloud ${CMAKE_PROJECT_NAME})
+TOOL(ManuallyCropGeometry   ${CMAKE_PROJECT_NAME})
```
Comment (syncle) [inline src/Tools/CMakeLists.txt]: I changed the application name. Do we have any dependent projects regarding this program names?

File: src/Visualization/Utility/SelectionPolygon.cpp [hunk 3]
```diff
@@ -182,6 +200,18 @@ std::shared_ptr<PointCloud> SelectionPolygon::CropPointCloudInPolygon(
...
+        const TriangleMesh &input, const ViewControl &view)
+{
+    return SelectDownSample(input, CropInRectangle(input.vertices_, view));
+}
+
+std::shared_ptr<TriangleMesh> SelectionPolygon::CropTriangleMeshInPolygon(
+        const TriangleMesh &input, const ViewControl &view)
+{
+    return SelectDownSample(input, CropInPolygon(input.vertices_, view));
+}
```
Comment (syncle) [inline src/Visualization/Utility/SelectionPolygon.cpp]: Can we use template functions?

File: src/Visualization/Utility/SelectionPolygon.h [hunk 4]
```diff
@@ -71,6 +74,10 @@ class SelectionPolygon : public Geometry2D
             const PointCloud &input, const ViewControl &view);
     std::shared_ptr<PointCloud> CropPointCloudInPolygon(
             const PointCloud &input, const ViewControl &view);
+    std::shared_ptr<TriangleMesh> CropTriangleMeshInRectangle(
+            const TriangleMesh &input, const ViewControl &view);
+    std::shared_ptr<TriangleMesh> CropTriangleMeshInPolygon(
+            const TriangleMesh &input, const ViewControl &view);
```
Comment (syncle) [inline src/Visualization/Utility/SelectionPolygon.h]: Same as previous comment, maybe template functions.

File: src/Visualization/Visualizer/VisualizerWithEditing.cpp [hunk 5]
```diff
@@ -392,6 +393,33 @@ void VisualizerWithEditing::KeyPressCallback(GLFWwindow *window,
...
+                    filename = default_filename.c_str();
+                }
+                if (filename == NULL) {
+                    PrintInfo("No filename is given. Abort saving.\n");
+                } else {
+                    SaveCroppingResult(filename);
+                }
+                view_control.ToggleLocking();
+                InvalidateSelectionPolygon();
+                InvalidatePicking();
```
Comment (syncle) [inline src/Visualization/Visualizer/VisualizerWithEditing.cpp]: I want to reduce these lines. This is too redundant and quite similar to previous lines in if statement. Maybe define template functions that does this job.

File: src/Tools/CMakeLists.txt (hunk 2 above)
Comment (qianyizh) [reply src/Tools/CMakeLists.txt]: I think this is fine.

File: src/Visualization/Utility/SelectionPolygon.cpp (hunk 3 above)
Comment (qianyizh) [reply src/Visualization/Utility/SelectionPolygon.cpp]: Yes, as long as the implementation gets declared in .cpp files and does not leak into .h.

File: src/Visualization/Utility/SelectionPolygon.h (hunk 4 above)
Comment (qianyizh) [reply src/Visualization/Utility/SelectionPolygon.h]: I think it is fine. Let's leave it for now.
Plus, I can't think of any other use cases. These two are all that we need.

File: src/Visualization/Visualizer/VisualizerWithEditing.cpp (hunk 5 above)
Comment (qianyizh) [reply src/Visualization/Visualizer/VisualizerWithEditing.cpp]: Up to you.

File: src/Core/Geometry/DownSample.cpp [hunk 6]
```diff
@@ -103,6 +104,74 @@ std::shared_ptr<PointCloud> SelectDownSample(const PointCloud &input,
...
+    // Only a face with three selected points contributes to mark mask_observed_vertex.
+    std::vector<bool> mask_observed_vertex(input.vertices_.size());
+    for (auto vertex_ids : triangle_to_vertex) {
+        if ((int)vertex_ids.size() == 3)
+            for (int i=0; i<3; i++)
+                mask_observed_vertex[vertex_ids[i]] = true;
+    }
+    // Rename vertex id based on selected points
+    std::vector<int> new_vertex_id(input.vertices_.size());
+    for (int i=0, cnt=0; i<mask_observed_vertex.size(); i++) {
```
Comment (qianyizh) [inline src/Core/Geometry/DownSample.cpp]: VS does not throw out warnings?
I think you should use ``i < (int)mask_observed_vertex.size()``, or ``auto i``

File: src/Core/Geometry/DownSample.cpp [hunk 7]
```diff
@@ -103,6 +104,74 @@ std::shared_ptr<PointCloud> SelectDownSample(const PointCloud &input,
...
+        if ((int)vertex_ids.size() == 3) {
+            Eigen::Vector3i new_face;
+            for (int i=0; i<3; i++)
+                new_face(i) = new_vertex_id[input.triangles_[triangle_id][i]];
+            output->triangles_.push_back(new_face);
+        }
+        triangle_id++;
+    }
+    // Push marked vertex.
+    for (int i=0; i<mask_observed_vertex.size(); i++) {
```
Comment (qianyizh) [inline src/Core/Geometry/DownSample.cpp]: Same comment as above.

File: src/Core/Geometry/DownSample.cpp [hunk 8]
```diff
@@ -103,6 +104,74 @@ std::shared_ptr<PointCloud> SelectDownSample(const PointCloud &input,
...
+    }
+    // Push a triangle that has 3 selected vertices.
+    triangle_id = 0;
+    for (auto vertex_ids : triangle_to_vertex) {
+        if ((int)vertex_ids.size() == 3) {
+            Eigen::Vector3i new_face;
+            for (int i=0; i<3; i++)
+                new_face(i) = new_vertex_id[input.triangles_[triangle_id][i]];
+            output->triangles_.push_back(new_face);
+        }
```
Comment (qianyizh) [inline src/Core/Geometry/DownSample.cpp]: triangle_normals_?

File: src/Visualization/Utility/SelectionPolygon.cpp [hunk 9]
```diff
@@ -127,6 +128,23 @@ std::shared_ptr<PointCloud> SelectionPolygon::CropPointCloud(
     }
 }
 
+std::shared_ptr<TriangleMesh> SelectionPolygon::CropTriangleMesh(
```
Comment (qianyizh) [inline src/Visualization/Utility/SelectionPolygon.cpp]: Should these functions be exposed in Python too?

File: src/Core/Geometry/DownSample.cpp [hunk 10]
```diff
@@ -103,6 +104,74 @@ std::shared_ptr<PointCloud> SelectDownSample(const PointCloud &input,
     return output;
 }
 
+std::shared_ptr<TriangleMesh> SelectDownSample(const TriangleMesh &input,
```
Comment (qianyizh) [inline src/Core/Geometry/DownSample.cpp]: should this be exposed in python too?

**Comment by syncle** [comment]
Thanks @qianyizh. I will revisit this once the other PR is addressed.

File: src/Visualization/Utility/SelectionPolygon.cpp (hunk 9 above)
Comment (syncle) [reply src/Visualization/Utility/SelectionPolygon.cpp]: done.

File: src/Core/Geometry/DownSample.cpp (hunk 10 above)
Comment (syncle) [reply src/Core/Geometry/DownSample.cpp]: done.

**Comment by qianyizh** [comment]
This is great.
I think this document should be edited too.
Add a note or something saying that meshes are also supported.
http://www.open3d.org/docs/tutorial/Basic/pointcloud.html#crop-point-cloud
But this is a minor fix and can be done later.
I will merge it.

**Pull request merged**